### PR TITLE
[export_python] inductor: emit only the preamble bindings a graph uses

### PR DESCRIPTION
**Pull request by bobrenjc93**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.17.0) (oldest at bottom):
* #193901
* #193900
* #193899
* #198257
* #198256
* #198255
* #198254
* #198253
* #198252
* #193898
* #198247
* #198246
* #198245
* __->__ #193974
* #193973

**Problem.** Under `readable_wrapper` the emitted module is the thing the user reads and edits, but its preamble imports and binds everything inductor might need on any backend, so a small CUDA softmax opens with ctypes, tempfile, random, XPU/MTIA allocators and a dozen other lines it never touches.

**Fix.** The preamble becomes a table of `(names bound, source line)` entries, and in readable mode each entry is emitted as a deferred line that is dropped at assembly time unless the finished module actually uses one of its names; `torch` is always kept.

**Summary.** Inductor writes the wrapper preamble in `write_header`, which runs from `__init__` before any node is lowered, so the preamble could not depend on what the graph needs and every graph got every backend's imports and bindings. In readable mode that made the preamble the bulk of a small artifact: a CUDA softmax was 182 lines and a CPU reduction 111. This commit splits the preamble into named entries (`_preamble_imports`, `_preamble_bindings`) routed through `write_preamble_line`, and `ReadablePythonWrapperCodegen` wraps each in a `_LineIfNamesUsed` that asks, at `getvalue()` time, whether any of its names appears in the finished module. The scan has to ignore three things that look like uses but are not -- a binding's own definition, comments, and hoisted kernel bodies -- and it also drops the inert autotuning block and the benchmark harness, bringing the two graphs to 134 and 68 lines. Default codegen is unchanged: `write_preamble_line` emits every line unconditionally, and the generated module is byte-identical to the two blobs it replaced.

## The preamble, before and after

The parent spliced two fixed blobs into every module. Here is that preamble for a CUDA softmax (`torch.softmax(x * 2, dim=-1)` on a `64x128` CUDA tensor, the fixture used by the tests), annotated with what readable mode now does with each line where a test pins it:

```python
from ctypes import c_void_p, c_long, c_int           # now: dropped
import torch                                         # now: always kept (_ALWAYS_EMIT)
import math                                          # now: dropped (only the kernel's `math as tl_math` used it)
import random                                        # now: dropped
import os
import tempfile                                      # now: dropped
from math import inf, nan
from cmath import nanj
from torch._inductor.hooks import run_intermediate_hooks   # now: dropped
from torch._inductor.utils import maybe_profile            # now: dropped
from torch._inductor.codegen.memory_planning import _align as align
from torch import device, empty_strided              # now: dropped (only kernel metadata said 'device': 0)
...
import triton                                        # now: dropped from the preamble (the hoisted kernel imports it)
import triton.language as tl                         # now: dropped from the preamble
from torch._inductor.runtime.triton_heuristics import start_graph, end_graph   # now: dropped

aten = torch.ops.aten                                # now: dropped (only comments named aten)
inductor_ops = torch.ops.inductor
_quantized = torch.ops._quantized                    # now: dropped
assert_size_stride = torch._C._dynamo.guards.assert_size_stride   # now: kept, the graph calls it
...
empty_strided_cpu_pinned = torch._C._dynamo.guards._empty_strided_cpu_pinned   # now: dropped
empty_strided_cuda = torch._C._dynamo.guards._empty_strided_cuda               # now: kept
empty_strided_xpu = torch._C._dynamo.guards._empty_strided_xpu                 # now: dropped
empty_strided_mtia = torch._C._dynamo.guards._empty_strided_mtia               # now: dropped
alloc_from_pool = torch.ops.inductor._alloc_from_pool                          # now: dropped
```

Lines without an annotation follow the same rule (kept if and only if the module references a name they bind); the tests just do not pin them individually. For a CPU graph (`(x + 1).relu().sum(0)` on a 1024-element tensor), `empty_strided_cpu` survives and `empty_strided_cuda` / `empty_strided_xpu` do not.

`torch` is the one exception: it is used by essentially every graph and is what the rest of the preamble is written in terms of (`aten = torch.ops.aten`, `empty_strided_cuda = torch._C...`), so dropping it could only ever be wrong:

```python
_ALWAYS_EMIT = ("torch",)
```

| graph (from the export decorator) | parent | now |
| --- | --- | --- |
| CUDA softmax | 182 lines | 134 lines |
| CPU reduction | 111 lines | 68 lines |

## Why decide at assembly time instead of at each use site

The alternative was to route the ~30 use sites through a registry that records "this graph needs X". That cannot be total, because several of these names are never spelled out at a use site:

```python
f"empty_strided_{device.type}"   # interpolated: no use site mentions empty_strided_cuda
repr(torch.device("cuda", 0))     # -> "device(type='cuda', index=0)"; uses `device`
repr(float("inf"))                # -> "inf"; uses `inf` / `nan`
repr(aten.scatter_.src)           # -> "aten.scatter_.src"; uses `aten`
```

So the question is asked once the module is finished. `write_preamble_line` is the hook: the base class writes the line as-is (which of them a graph will use is not known when `write_header` runs), and the readable wrapper writes a deferred line instead:

```python
# PythonWrapperCodegen
def write_preamble_line(self, buf, names, line):
    buf.writeline(line)

# ReadablePythonWrapperCodegen
def write_preamble_line(self, buf, names, line):
    if any(name in _ALWAYS_EMIT for name in names):
        buf.writeline(line)
        return
    buf.writeline(_LineIfNamesUsed(line, names, self))
```

`_LineIfNamesUsed.__call__` returns the line only if `preamble_names_used(names)` finds one of the names as a whole word (`\bname\b`) in a scan text built lazily, once, from every buffer that can hold a use: `imports`, `header` (which by then also holds the replayed kernel definitions), `subgraph_definitions`, `prefix`, `wrapper_call`, `suffix` and `kernel_declarations`. The `empty_strided_p2p` binding, emitted only when `_SymmetricMemory` is importable, goes through the same hook.

## Three things that look like uses and are not

Each of these is a way a naive scan gets it wrong, and each has a test.

**A binding must not be kept alive by its own definition.** `_quantized = torch.ops._quantized` names itself on the right-hand side, so counting attribute names as uses makes it permanently unprunable. While the scan text is being built, `scanning_for_uses` is set and every `_LineIfNamesUsed` returns `None`, so every preamble line reads as absent and no definition can witness itself.

```python
# a naive scan: "_quantized" in text -> True, line kept forever
# now: the preamble is invisible to its own scan; _quantized does not appear in the module
```

Pinned by `test_a_binding_is_not_kept_alive_by_its_own_definition`.

**A name in a comment is not a use.** Inductor stamps each kernel with a provenance comment listing its source ops, so a comment-blind scan concludes every graph uses `aten`:

```python
# Original ATen: [aten.mul, ...]
```

Lines whose first non-blank character is `#` are removed from the scan text. `test_a_name_mentioned_only_in_a_comment_is_not_a_use` checks that the module still contains `Original ATen: [aten.` but not `aten = torch.ops.aten`.

**A name inside a kernel is not a wrapper use.** A kernel module is self-contained -- that is exactly why the default path can exec each one in its own namespace -- and it supplies its own imports and metadata:

```python
import triton
import triton.language as tl
from torch._inductor.runtime.triton_helpers import libdevice, math as tl_math
...
triton_meta={..., 'device': 0, ...}
```

Once kernels are hoisted they share the wrapper's text, so the readable wrapper overrides `_define_kernel_helper` to record each `kernel_body`, and `preamble_names_used` removes those texts before scanning. `test_a_name_used_only_inside_a_kernel_is_not_a_wrapper_use` checks that `math as tl_math` is present while `\nimport math\n` and `from torch import device, empty_strided` are not.

The same reasoning covers the triton header: `write_triton_header_once` in readable mode emits `import triton`, `import triton.language as tl` and `start_graph, end_graph` (only used under `profile_bandwidth`) through `write_preamble_line`, so the hoisted kernel's own imports are the only copy. `test_triton_is_not_imported_twice` asserts that neither `import triton` nor `start_graph` is in the preamble and that each of `import triton\n` and `import triton.language as tl\n` appears exactly once in the module. When `autotune_at_compile_time` or `cpp_wrapper` is on, the parent's `write_triton_header_once` runs unchanged, since those paths splice into buffers this mode does not prune.

## Also dropped in readable mode

```python
# graph.py, _compile_to_module_lines
if config.triton.autotune_at_compile_time and not config.readable_wrapper:
    ...  # embed kernel_autotune_defs as an inert string
```

The compile-time autotuning block is a record of code that was exec'd at compile time, carried in the module as an inert string for debugging, and it was the single largest thing in a small artifact.

`add_benchmark_harness` is a no-op in readable mode. The harness (`get_args` / `benchmark_compiled_module` / `__main__`) is a runnable benchmark script bolted onto a module this mode wants read rather than run, and it is written into the output buffer after the preamble has been scanned, so it could reference a binding already dropped.

## Default codegen is untouched

With the default wrapper, `write_preamble_line` emits every entry, and the preamble table emits exactly the two blobs it replaced, verified by diffing the generated module byte for byte. `test_preamble_binds_only_what_the_graph_uses` pins the positive side in readable mode: the surviving preamble still contains `empty_strided_cuda` and `assert_size_stride`, which the graph does use.

Test Plan:

```
python test/inductor/test_readable_wrapper.py
python test/inductor/test_wrapper_codegen.py
python test/inductor/test_compile_to_python.py
python test/inductor/test_torchinductor.py
```

6 new tests: one per failure mode above, one that a CPU graph binds no GPU
allocator, one that the surviving preamble still contains what the graph does
use, and one that `import triton` appears exactly once now that a hoisted kernel
supplies its own.

This commit was authored with the assistance of an AI coding agent.

### CODE REVIEW

**Comment by bobrenjc93** [comment]
@claude review this PR

**Comment by bobrenjc93** [comment]
Responses to the review above, one per finding. This PR now also owns the "emit the AsyncCompile lifecycle only when the finished module uses it" decision that moved here from #193973 (see the reply there), decided by the same whole-module scan.

**1. Subgraph kernel bodies not excluded from the parent's scan**
Fixed in b68e80c37a0: the readable subgraph wrapper shares the root's `_kernel_texts` list (mirroring `src_to_kernel`). Covered by the cond-graph subTest of `test_a_name_used_only_inside_a_kernel_is_not_a_wrapper_use`; un-sharing the list fails it.

**2. `write_triton_header_once` override shadows the subgraph delegation (mid-file `import triton`)**
Fixed in b68e80c37a0: the readable override is removed. The base method now routes its lines through `write_preamble_line`, so `SubgraphPythonWrapperCodegen`'s delegation to the parent wins by MRO. The per-subgraph `get_raw_stream` line is identical with readable off, so it is unchanged here. `test_triton_is_not_imported_twice` now also runs on a cond graph.

**3. `_ALWAYS_EMIT` matched per-name**
Declined: with the current table this is exactly right. The concern is about a hypothetical future entry.

**4. `profile_bandwidth_output` hits AttributeError**
Fixed in b68e80c37a0: `readable_wrapper_requested()` refuses the `profile_bandwidth_output` combination alongside `benchmark_kernel`, with a test.

**5. Tests tautological on single-kernel softmax; fragile split key**
Fixed in b68e80c37a0: both properties are also asserted on the cond / partitioned graph. `test_cpu_graph_does_not_bind_gpu_allocators` forces the cpp backend and asserts on the whole module instead of splitting on `async_compile.cpp`.

**Kernel-text subtraction can silently no-op after `splice` reformatting**
Fixed in b68e80c37a0: the recorded kernel text is the post-splice header lines, recorded only for hoisted kernels.

**6. O(entries x text) regex scan**
Fixed in b68e80c37a0: the module is tokenized once into a set of identifiers, then there is one `isdisjoint` check per entry.

**Comment by bobrenjc93** [comment]
@claude review this PR

**Comment by bobrenjc93** [comment]
@claude review this PR

**Comment by bobrenjc93** [comment]
Response to the previous review:

**1. The description claims a `graph.py` change, `_ALWAYS_EMIT = ("torch",)` and a line-count table that are not in this commit.** Fixed. The PR description was rebuilt from this commit alone; the `graph.py` gate and the AsyncCompile lifecycle live in #198274. After #4 below, `_ALWAYS_EMIT` is now `("torch",)` in this commit.

**2. Scan the assembled `result` instead of a hand-listed buffer list.** Declined for this PR. It would change what the scan is and would reopen the harness and `profile_bandwidth_output` behavior, which is more than this PR should take on. As suggested, the invariant is now stated next to the buffer list in ad570b5e738: `generate()` writes a few lines straight into `result`, and none of them may name a preamble binding.

**3. Only whole-line comments are excluded.** Fixed in ad570b5e738. The comment now says exactly that: whole-line comments are skipped, while trailing comments and string literals still count, which can only keep a line, never drop a needed one.

**4. `async_compile` in `_ALWAYS_EMIT` is dead.** Fixed in ad570b5e738. It is dropped, since the wait/del in `prefix` keeps it alive through the scan. The `AsyncCompile` import entry now carries a comment saying it is keyed by the binding it exists for, so the two are coupled deliberately.

**5. No test that default codegen emits every table entry.** Fixed in ad570b5e738. `test_default_wrapper_emits_the_whole_preamble` spies on `write_preamble_line` and asserts that every entry of `_preamble_imports()` and `_preamble_bindings()` reaches the default wrapper's output. I checked that it fails when the default path drops one entry.

**Comment by bobrenjc93** [comment]
@claude review this PR

**Comment by bobrenjc93** [comment]
Response to the previous review:

**1. Nothing pins the direction that breaks the artifact.** Fixed in 41eff387904. Added `test_names_the_wrapper_uses_are_kept`. The graph's wrapper code names `device(...)` through an extern `randperm(device=...)` and `inf` through `_cdist_forward(..., inf, None)`. The test asserts that both `from torch import device, empty_strided` and `from math import inf, nan` survive, and that the module runs standalone and matches the compiled result. Forcing the scan to report either name as unused fails it.

**2. The scan counts tokens inside string literals.** Declined. That is the safe direction: a false use keeps a binding, and it never drops one the module needs. The kernels that stay inline (user `@triton.jit`, C++, MPS, Halide, Pallas) are out of this PR's scope. The scan text is concatenated partial buffers, so it is not guaranteed to parse, and a tokenizer or `ast` pass would add machinery for a size-only gain. Tightening this is a reasonable follow-up.

**3. `standalone` read out of `**kwargs`.** Fixed in 41eff387904. The override now mirrors the base signature exactly, so `standalone` is positional-or-keyword as the 6th parameter and is forwarded positionally. It also skips recording an empty slice.

**Comment by bobrenjc93** [comment]
@claude review this PR

**Comment by bobrenjc93** [comment]
Response to the round-8 review (AI-assisted, drafted by Claude):

> **1. `benchmark_harness` silently ignored**

Declined. `benchmark_harness` defaults to True, so refusing it would make `readable_wrapper=True` fail out of the box unless every caller also disabled the harness. Dropping the harness is a documented property of the mode: the commit message says so, and `profile_bandwidth_output`, the in-tree consumer that needs the harness, is refused. The readable module is meant to be read and imported, not run as a benchmark script.

> **2. Subgraph docstring no longer describes the class**

Fixed in 31bd7799b5d. The docstring now lists only the subgraph overrides the readable class leaves alone, and says that both classes drop the benchmark harness.

> **3. `__init__` signatures**

Declined, since it's a nit. `create()` is the only constructor path.

> **4. Coverage of the triton trio's default shape / ordering / cpp tokens in the scan**

Declined. Default-path equivalence was verified by a byte-for-byte diff of the generated module for a CPU and a CUDA graph. A C++ kernel string being in the scan is what keeps `async_compile` alive, and `test_async_compile_survives_when_a_backend_needs_it` one PR up asserts exactly that.

**Comment by bobrenjc93** [comment]
@claude review this PR
